### PR TITLE
[SMD-82]: authorisation

### PR DESCRIPTION
**Pull request by Joel-Pearce**
Authorisation of local authorities: https://dluhcdigital.atlassian.net/jira/software/c/projects/SMD/boards/140?selectedIssue=SMD-82

Derives the LA from user's e-mail, and uses that to look up a list of place names the user is allowed to submit returns for. This list is sent to data store, where the authorisation check will occur.

- [ ] Unit tests and other appropriate tests added or updated
- [ ] README and other documentation has been updated / added (if needed)
- [ ] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### CODE REVIEW

File: app/main/authorisation.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    authority is found, its identifier is returned, and this is used to look up the
+    place names for which the local authority is expected to submit for.
+
+    :param user_email: A string representing the user's email address.
+    :return: An array of place names or None if none are found.
+    """
+    lookup_url = Config.LOCAL_AUTHORITIES_LOOKUP
+    df_lookup = pd.read_csv(lookup_url, index_col=0)
+
+    organisation_id = df_lookup[df_lookup["homepage_url"].str[11:-1] == user_email.split("@")[1]]["gss"]
```
Comment (Connor-OS) [inline app/main/authorisation.py:22]: is .str[11:-1] intended to remove the https protocal from the front the url. If so to deal with http(s)

i.e:
https://www.monmouthshire.gov.uk/
http://www.moray.gov.uk

could we split on . instead?

Comment (Joel-Pearce) [reply app/main/authorisation.py:22]: Yes, we might be able to. I'll have a look at the list and see if there's any exceptions, and if not re-implement by splitting on the "." char and selecting the right portion of the array if we decide to continue with this approach of authorising via e-mail domain.

Comment (Joel-Pearce) [reply app/main/authorisation.py:22]: Thinking about it, we can't split on the "." char because some urls have a "/" at the end and some don't, which is how I arrived at the logic for 11:-1 vs. 11:, but we can just generically remove the https protocol and any trailing slashes by default and then do the logic on the dataframe lookup. That will save us from having to do the lookup logic twice over.

File: app/main/authorisation.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    # some urls do not end with "/"
+    if len(organisation_id) == 0:
+        organisation_id = df_lookup[df_lookup["homepage_url"].str[11:] == user_email.split("@")[1]]["gss"]
+
+    if len(organisation_id) == 0:
```
Comment (gidsg) [inline app/main/authorisation.py:28]: Could we maybe use a regex instead to capture both styles? As is it makes the logic hard to parse? Or maybe use a library like https://pypi.org/project/tld/ ?

File: app/main/authorisation.py (hunk 1 above)
Comment (gidsg) [reply app/main/authorisation.py:22]: Could we use an external library, as per other comment, or use `netloc` from `urlparse` in the core library, which should give us a more standardised string?

File: requirements.in [hunk 3]
```diff
@@ -6,3 +6,4 @@ flask-assets==2.0
 govuk-frontend-jinja==2.7.0
 gunicorn==20.1.0
 bs4
+pandas
```
Comment (cyrusdobbs) [inline requirements.in:9]: Can this be removed now?

Comment (Joel-Pearce) [reply requirements.in:9]: Ah, yes, it can, I'll remove it.

File: app/const.py [hunk 4]
```diff
@@ -4,3 +4,130 @@
...
+    "wiltshire.gov.uk": ["Trowbridge", "Salisbury City Centre"],
+    "wirral.gov.uk": ["New Ferry", "Birkenhead"],
+    "wolverhampton.gov.uk": ["Wolverhampton", "Wolverhampton City Centre (West)"],
+    "worcester.gov.uk": ["Blackfriars - Northern City Centre", "Worcester"],
+    "wyreforestdc.gov.uk": ["Kidderminster"],
+    "bcpcouncil.gov.uk": ["Bournemouth"],
+    "eastsuffolk.gov.uk": ["Lowestoft"],
+    "northnorthants.gov.uk": ["Corby"],
+    "westnorthants.gov.uk": ["Northampton"],
+}
```
Comment (cyrusdobbs) [inline app/const.py:133]: Could these lists be converted to tuples as they're a more efficient data structure to use if we don't care about their immutability

Comment (Joel-Pearce) [reply app/const.py]: Yes, they're now tuples

**Review by cyrusdobbs — APPROVED** [review]
LGTM 🥥
